[WebXR] Use MTLTexture to attach color IOSurface

djg opened the pull request:
#### d41c610090b07554b2266f6d56435c8612c987b4
<pre>
[WebXR] Use MTLTexture to attach color IOSurface
<a href="https://bugs.webkit.org/show_bug.cgi?id=258598">https://bugs.webkit.org/show_bug.cgi?id=258598</a>
rdar://problem/111426880

Reviewed by Kimmo Kinnunen.

There are two paths in WebXROpaqueFramebuffer for attaching compositor provided
IOSurface for rendering: 1) Via MTLSharedTexture, and 2) Via pbuffer.

Since it&apos;s possible to create a MTLTexture from an IOSurface and plane index,
switch away from using pbuffer to reduce the number of code paths.

This patch introduces USE(MTLTEXTURE_FOR_XR_LAYER_DATA) to aid the transition of
platform specific WebXR systems. Once those systems are updated,
USE(IOSURFACE_FOR_XR_LAYER_DATA) will be removed.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::allocateColorStorage):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::platformInitialize):
(WebCore::GraphicsContextGLCocoa::createAndBindEGLImage):
(WebCore::createAndBindSharedTextureToExternalImage): Deleted.
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::FrameData::LayerData::encode const):
(PlatformXR::Device::FrameData::LayerData::decode):
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):

Canonical link: <a href="https://commits.webkit.org/265606@main">https://commits.webkit.org/265606@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/e740e59f3448aeae3f353f14234fac8cae0c8b42

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/11384 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/11589 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/11935 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/13029 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/10836 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/11405 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/13971 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/11571 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/13754 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/11546 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/12416 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/9631 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/13448 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/9708 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/10319 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/17497 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/10779 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/10473 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/13682 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/10889 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/8961 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/10062 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/2725 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/14335 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/10744 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

pxlcoder commented on Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
```diff
@@ -683,85 +683,67 @@ static bool needsEAGLOnMac()
-}
-#endif
+    GCGLuint textureWidth = [texture.get() width];
+    GCGLuint textureHeight = [texture.get() height];
```
`.get()` is unnecessary.

kkinnunen-apple commented on Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
```diff
@@ -35,6 +35,10 @@
 #include <wtf/RetainPtr.h>
 
+#if USE(IOSURFACE_FOR_XR_LAYER_DATA) || USE(MTLTEXTURE_FOR_XR_LAYER_DATA)
+#define USE_WEBXR_USE_EGL_IMAGE 1
```
It's a bit inconvenient to signal that "this is a standard USE() macro" but then at the same time not have it in standard place, available for all .cpp files, and at the same time not having it confied to just one .cpp file. 
So maybe this could be just omitted and written inline like
`#if USE(IOSURFACE_FOR_XR_LAYER_DATA) || USE(MTLTEXTURE_FOR_XR_LAYER_DATA)`.

There's 4 uses of USE_WEBXR_USE_EGL_IMAGE So if you remove the definition, you have penalty of score 3.

If I understand correctly, the target is to replace
`USE(IOSURFACE_FOR_XR_LAYER_DATA) || USE(MTLTEXTURE_FOR_XR_LAYER_DATA`
with
`PLATFORM(COCOA)`..

djg committed:
[WebXR] Use MTLTexture to attach color IOSurface
https://bugs.webkit.org/show_bug.cgi?id=258598
rdar://problem/111426880

Reviewed by Kimmo Kinnunen.

There are two paths in WebXROpaqueFramebuffer for attaching compositor provided
IOSurface for rendering: 1) Via MTLSharedTexture, and 2) Via pbuffer.

Since it's possible to create a MTLTexture from an IOSurface and plane index,
switch away from using pbuffer to reduce the number of code paths.

This patch introduces USE(MTLTEXTURE_FOR_XR_LAYER_DATA) to aid the transition of
platform specific WebXR systems. Once those systems are updated,
USE(IOSURFACE_FOR_XR_LAYER_DATA) will be removed.

* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.cpp:
(WebCore::WebXROpaqueFramebuffer::startFrame):
(WebCore::WebXROpaqueFramebuffer::endFrame):
(WebCore::WebXROpaqueFramebuffer::allocateColorStorage):
* Source/WebCore/Modules/webxr/WebXROpaqueFramebuffer.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.h:
* Source/WebCore/platform/graphics/cocoa/GraphicsContextGLCocoa.mm:
(WebCore::GraphicsContextGLCocoa::platformInitialize):
(WebCore::GraphicsContextGLCocoa::createAndBindEGLImage):
(WebCore::createAndBindSharedTextureToExternalImage): Deleted.
* Source/WebCore/platform/xr/PlatformXR.h:
(PlatformXR::Device::FrameData::LayerData::encode const):
(PlatformXR::Device::FrameData::LayerData::decode):
* Source/WebCore/testing/WebFakeXRDevice.cpp:
(WebCore::SimulatedXRDevice::frameTimerFired):

Canonical link: https://commits.webkit.org/265606@main
